File: 滑动窗口/不定长滑动窗口/最多k个重复元素的最长子数组.py
from collections import defaultdict 

def test(nums: list[int], k: int):
    dic = defaultdict(int)
    tmp_res = 0
    res = 0
    left = 0
    for i, num in enumerate(nums):

        while dic[num] >= k:
            dic[nums[left]] -= 1
            tmp_res -= 1
            left += 1

        dic[num] += 1
        tmp_res += 1
        res = max(tmp_res, res)
    return res

nums = [1,2,3,1,2,3,1,2]
print(test(nums, 2))

File: 滑动窗口/不定长滑动窗口/最多k个重复元素的最长子数组.cpp
#include <iostream>
#include <vector>
#include <map>
using namespace std;

class Solution{
    public:
    int test(vector<int>& nums, int k){
        map<int, int> mp;
        int tem_res = 0;
        int res = 0;
        int left = 0;
        for(int i = 0; i < nums.size(); i++){
            while(mp[nums[i]] >= k){
                mp[nums[left]] -= 1;
                tem_res -= 1;
                left ++;
            }
            mp[nums[i]] += 1;
            tem_res += 1;
            res = max(res, tem_res);

        }
        return res;

    }
};

int main(){
    Solution solution;
    vector<int> nums = {1,2,3,1,2,3,1,2};
    int k = 2;
    cout << "Test1: " << solution.test(nums, k) << endl;
}
